In [1]:
import numpy as np
import pandas as pd
import datetime as dt

In [2]:
df=pd.read_csv(r'C:\Users\Suvakanta\Desktop\DS_Internship_2023\Data Analysis\diminos_data.csv')
df

order_id      order_placed_at          order_delivered_at
0       1523111  2023-03-01 00:00:59  2023-03-01 00:18:07.443132
1       1523112  2023-03-01 00:03:59  2023-03-01 00:19:34.925241
2       1523113  2023-03-01 00:07:22  2023-03-01 00:22:28.291385
3       1523114  2023-03-01 00:07:47  2023-03-01 00:46:19.019399
4       1523115  2023-03-01 00:09:03  2023-03-01 00:25:13.619056
...         ...                  ...                         ...
14995   1538106  2023-03-27 23:37:05  2023-03-27 23:52:37.409378
14996   1538107  2023-03-27 23:47:38  2023-03-28 00:04:22.672912
14997   1538108  2023-03-27 23:50:16  2023-03-28 00:05:40.676238
14998   1538109  2023-03-27 23:52:44  2023-03-28 00:08:41.810358
14999   1538110  2023-03-27 23:58:20  2023-03-28 00:13:42.499311

[15000 rows x 3 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 3 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            15000 non-null  int64 
 1   order_placed_at     15000 non-null  object
 2   order_delivered_at  15000 non-null  object
dtypes: int64(1), object(2)
memory usage: 351.7+ KB


In [4]:
df.isnull().sum()

order_id              0
order_placed_at       0
order_delivered_at    0
dtype: int64

In [5]:
df['order_placed_date']=df['order_placed_at'].apply(lambda x:x.split()[0])
df['order_placed_time']=df['order_placed_at'].apply(lambda x:x.split()[1])
df['order_delivered_date']=df['order_delivered_at'].apply(lambda x:x.split()[0])
df['order_delivered_time']=df['order_delivered_at'].apply(lambda x:x.split()[1])

In [6]:
df.drop(columns=['order_placed_at','order_delivered_at'],inplace=True)

In [7]:
df.head()

order_id order_placed_date order_placed_time order_delivered_date  \
0   1523111        2023-03-01          00:00:59           2023-03-01   
1   1523112        2023-03-01          00:03:59           2023-03-01   
2   1523113        2023-03-01          00:07:22           2023-03-01   
3   1523114        2023-03-01          00:07:47           2023-03-01   
4   1523115        2023-03-01          00:09:03           2023-03-01   

  order_delivered_time  
0      00:18:07.443132  
1      00:19:34.925241  
2      00:22:28.291385  
3      00:46:19.019399  
4      00:25:13.619056

In [8]:
#df['order_place_date']=df['order_placed_date'].apply(lambda x:x.split("-")[2])
#df['order_placed_month']=df['order_placed_date'].apply(lambda x:x.split("-")[1])
#df['order_placed_year']=df['order_placed_date'].apply(lambda x:x.split("-")[0])

#df['order_deliver_date']=df['order_delivered_date'].apply(lambda x:x.split("-")[2])
#df['order_delivered_month']=df['order_delivered_date'].apply(lambda x:x.split("-")[1])
#df['order_delivered_year']=df['order_delivered_date'].apply(lambda x:x.split("-")[0])

#df.drop(columns=['order_placed_date','order_delivered_date'],inplace=True)

In [9]:
#df['order_place_date']=df['order_place_date'].astype(int)
#df['order_placed_month']=df['order_placed_month'].astype(int)
#df['order_placed_year']=df['order_placed_year'].astype(int)

#df['order_deliver_date']=df['order_deliver_date'].astype(int)
#df['order_delivered_month']=df['order_delivered_month'].astype(int)
#df['order_delivered_year']=df['order_delivered_year'].astype(int)

In [10]:
def minute_placed(timestamp):
    hours, minutes, seconds = timestamp.split(":")
    minutes = int(minutes)
    seconds = int(seconds)
    total_minutes = minutes + (seconds / 60)
    return np.round(total_minutes,2)

df['order_placed_time(in minutes)']=df['order_placed_time'].apply(minute_placed)

In [11]:
def minute_delivered(timestamp):
    hours, minutes, seconds = timestamp.split(":")
    minutes = int(minutes)
    seconds = int(seconds.split('.')[0])
    total_minutes = minutes + (seconds / 60)
    return np.round(total_minutes,2)

df['order_delivered_time(in minutes)']=df['order_delivered_time'].apply(minute_delivered)

In [12]:
df.drop(columns=['order_placed_time','order_delivered_time'],inplace=True)

In [13]:
df.head()

order_id order_placed_date order_delivered_date  \
0   1523111        2023-03-01           2023-03-01   
1   1523112        2023-03-01           2023-03-01   
2   1523113        2023-03-01           2023-03-01   
3   1523114        2023-03-01           2023-03-01   
4   1523115        2023-03-01           2023-03-01   

   order_placed_time(in minutes)  order_delivered_time(in minutes)  
0                           0.98                             18.12  
1                           3.98                             19.57  
2                           7.37                             22.47  
3                           7.78                             46.32  
4                           9.05                             25.22

In [14]:
df.describe()

order_id  order_placed_time(in minutes)  \
count  1.500000e+04                   15000.000000   
mean   1.530610e+06                      29.818271   
std    4.330271e+03                      17.312219   
min    1.523111e+06                       0.000000   
25%    1.526861e+06                      14.865000   
50%    1.530610e+06                      29.710000   
75%    1.534360e+06                      44.585000   
max    1.538110e+06                      59.980000   

       order_delivered_time(in minutes)  
count                      15000.000000  
mean                          30.049294  
std                           17.256351  
min                            0.000000  
25%                           15.130000  
50%                           30.150000  
75%                           44.930000  
max                           59.980000

In [16]:
df['order_placed_date'] = pd.to_datetime(df['order_placed_date'])
df['order_delivered_date'] = pd.to_datetime(df['order_delivered_date'])

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15000 entries, 0 to 14999
Data columns (total 5 columns):
 #   Column                            Non-Null Count  Dtype         
---  ------                            --------------  -----         
 0   order_id                          15000 non-null  int64         
 1   order_placed_date                 15000 non-null  datetime64[ns]
 2   order_delivered_date              15000 non-null  datetime64[ns]
 3   order_placed_time(in minutes)     15000 non-null  float64       
 4   order_delivered_time(in minutes)  15000 non-null  float64       
dtypes: datetime64[ns](2), float64(2), int64(1)
memory usage: 586.1 KB


In [51]:
print('Maximum order per day :',df['order_placed_date'].dt.day.value_counts().max())
print()
print('Minimum order per day :',df['order_placed_date'].dt.day.value_counts().min())
print()
print('Average order per day :',np.round(df['order_placed_date'].dt.day.value_counts().mean()))

Maximum order per day : 597

Minimum order per day : 492

Average order per day : 556.0


In [40]:
df['Total Delivery time']=df['order_delivered_time(in minutes)'] - df['order_placed_time(in minutes)']

In [41]:
df.head()

order_id order_placed_date order_delivered_date  \
0   1523111        2023-03-01           2023-03-01   
1   1523112        2023-03-01           2023-03-01   
2   1523113        2023-03-01           2023-03-01   
3   1523114        2023-03-01           2023-03-01   
4   1523115        2023-03-01           2023-03-01   

   order_placed_time(in minutes)  order_delivered_time(in minutes)  \
0                           0.98                             18.12   
1                           3.98                             19.57   
2                           7.37                             22.47   
3                           7.78                             46.32   
4                           9.05                             25.22   

   Total Delivery time  
0                17.14  
1                15.59  
2                15.10  
3                38.54  
4                16.17

In [43]:
print('Maximum delivery time :',df['Total Delivery time'].max())
print()
print('Minimum delivery time :',df['Total Delivery time'].min())
print()
print('Average delivery time :',df['Total Delivery time'].median())

Maximum delivery time : 55.31

Minimum delivery time : -56.14

Average delivery time : 15.310000000000002


In [49]:
np.percentile(df['Total Delivery time'],95)

21.08049999999999

In [50]:
print((df[df['Total Delivery time']<=31.00]['Total Delivery time'].count())/len(df)*100)
print()
print((df[df['Total Delivery time']>31.00]['Total Delivery time'].count())/len(df)*100)

98.99333333333334

1.0066666666666666


In [77]:
print('Maximum order placed on :',df['order_placed_date'].value_counts().sort_values(ascending=False).index[0])
print()
print('Minimum order placed on :',df['order_placed_date'].value_counts().sort_values(ascending=False).index[-1])

Maximum order placed on : 2023-03-20 00:00:00

Minimum order placed on : 2023-03-14 00:00:00


### Observation -->
1. 98% of deliveries are within 31 minutes.
2. Average Delivery Time is 15 minutes, Maximum and minimum deliver time is 55 and 0 minutes respectively.
3. 95 percentile of orders is 21 minutes.
4. Maximum order per day : 597 & Minimum order per day : 492 & Average order per day : 556
5. Maximum order placed on : 2023-03-20 & Minimum order placed on : 2023-03-14